bmap-rs/src/main.rs

```diff
@@ -74,53 +83,150 @@ impl SeekForward for Decoder {
+            let mut xml = String::new();
+            b.read_to_string(&mut xml)?;
+            Bmap::from_xml(&xml)?
+        }
 
-    if output.metadata()?.is_file() {
-        ftruncate(output.as_raw_fd(), bmap.image_size() as i64)
-            .context("Failed to truncate file")?;
+        Input::Remote(_) => {
+            let url_bmap = bmap
+                .to_str()
+                .expect("Fail to convert remote path to str")
+                .parse::<Uri>()
+                .expect("Fail to convert remote path to url");
+            let mut bmap_res = match url_bmap.scheme_str() {
+                Some("https") => fetch_url_https(url_bmap).await?,
+                Some("http") => fetch_url_http(url_bmap).await?,
+                _ => {
+                    bail!("url is not http or https")
+                }
+            };
+            let mut c = Cursor::new(Vec::new());
+            while let Some(next) = bmap_res.data().await {
+                let chunk = next?;
+                c.write_all(&chunk)?;
+            }
+            c.set_position(0);
+            let mut xml = String::new();
+            c.read_to_string(&mut xml)?;
+            Bmap::from_xml(&xml)?
+        }
+    };
+
+    match input_type {
+        Input::Local(mut input) => {
+            let mut output = std::fs::OpenOptions::new()
+                .write(true)
+                .create(true)
+                .open(c.dest)?;
+
+            if output.metadata()?.is_file() {
+                ftruncate(output.as_raw_fd(), bmap.image_size() as i64)
+                    .context("Failed to truncate file")?;
+            }
+            bmap::copy(&mut input, &mut output, &bmap)?;
+            println!("Done: Syncing...");
+            output.sync_all().expect("Sync failure");
+        }
+        Input::Remote(res) => {
+            let mut output = tokio::fs::OpenOptions::new()
+                .write(true)
+                .create(true)
+                .open(c.dest)
+                .await?;
+            ftruncate(output.as_raw_fd(), bmap.image_size() as i64)
+                .context("Failed to truncate file")?;
+            let mut chunk = match c.image.extension().and_then(OsStr::to_str) {
+                Some("gz") => {
+                    let stream = res.into_body().into_stream();
+                    let stream = stream.map(|result| {
+                        result.map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))
+                    });
+                    let read = StreamReader::new(stream);
+                    let reader = async_compression::tokio::bufread::GzipDecoder::new(read);
+                    AsyncDiscarder::new(reader)
+                }
+                _ => bail!("Image file format not implemented"),
+            };
+            bmap::copy_async(&mut chunk, &mut output, &bmap).await?;
+            println!("Done: Syncing...");
+            output.sync_all().await.expect("Sync failure");
```

Comment:
This should probably also be generic ; There is no reason why local files should be sync and remote files async, they can be treated in the same way.